src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +602,231 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
   iovec plaintext_vec = {(void*)plaintext, plaintext_size};
   iovec record_vec = {(void*)record, record_allocated_size};
   return s2a_write_tls13_record(crypter, SSL3_RT_APPLICATION_DATA,
-                                &plaintext_vec, 1, record_vec, record_size,
-                                error_details);
+                                &plaintext_vec, /** plaintext_vec length **/ 1,
+                                record_vec, record_size, error_details);
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        size_t record_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting a TLS 1.3 record of size
+   *  |record_size|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  GPR_ASSERT(record_size >= SSL3_RT_HEADER_LENGTH + tag_size);
+  *plaintext_size = record_size - SSL3_RT_HEADER_LENGTH - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        const iovec* protected_vec,
+                                        size_t protected_vec_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting the TLS 1.3 payload stored in
+   *  |protected_vec|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  size_t total_vec_length = get_total_length(protected_vec, protected_vec_size);
+  GPR_ASSERT(total_vec_length >= tag_size);
+  *plaintext_size = total_vec_length - tag_size;
```

Comment:
Since this is another overload of s2a_max_plaintext_size and size_t record_size / (const iovec* protected_vec + size_t protected_vec_size) are unrelated to TLS, I would have expected a - SSL3_RT_HEADER_LENGTH here. Maybe ignore my comments about merging the two functions and we just need better names. We can discuss offline.

Reply:
This method has now been removed.